### PR TITLE
feat(Core/SetCurrentTrainer): Add multitrainer function

### PR DESCRIPTION
**Pull request by yangyitzzg**
use multifunctional trainers in modules

## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [x] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:
add this code to modules:
```c++
class multiTrainerScript : public CreatureScript
{
public:
    multiTrainerScript() : CreatureScript("multiTrainerScript") { }

    bool OnGossipHello(Player* player, Creature* creature) override
    {
        ClearGossipMenuFor(player);
        AddGossipItemFor(player, GOSSIP_ICON_VENDOR, "Learn My Spell", GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INN_INFO + 1);
        AddGossipItemFor(player, GOSSIP_ICON_VENDOR, "Bye Bye!", GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INN_INFO + 2);
        SendGossipMenuFor(player,DEFAULT_GOSSIP_MESSAGE, creature->GetGUID());
        return true;
    }

    bool OnGossipSelect(Player* player, Creature* creature, uint32 sender, uint32 action)
    {
        if (action == GOSSIP_ACTION_INN_INFO + 1)
        {
            switch (player->getClass())
            {
            case CLASS_WARRIOR:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 914);
                break;
            case CLASS_PALADIN:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 5492);
                break;
            case CLASS_HUNTER:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 5516);
                break;
            case CLASS_ROGUE:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 918);
                break;
            case CLASS_PRIEST:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 5484);
                break;
            case CLASS_DEATH_KNIGHT:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 29196);
                break;
            case CLASS_SHAMAN:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 20407);
                break;
            case CLASS_MAGE:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 331);
                break;
            case CLASS_WARLOCK:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 5496);
                break;
            case CLASS_DRUID:
                player->GetSession()->SendTrainerList(creature->GetGUID(), 5504);
                break;
            }
        }

        if (action == GOSSIP_ACTION_INN_INFO + 2)
        {
            CloseGossipMenuFor(player);
        }

        return true;
    }
};
``` 

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by yangyitzzg** [comment]
Add some content:
creature_template:trainer_type Must be set to be greater than 3 and add UNIT_NPC_FLAG_TRAINER

**Comment by IntelligentQuantum** [comment]
https://github.com/Rochet2/TrinityCore/compare/TrinityCore:3.3.5...multitrainer_3.3.5.diff

**Comment by yangyitzzg** [comment]
> https://github.com/Rochet2/TrinityCore/compare/TrinityCore:3.3.5...multitrainer_3.3.5.diff

I'm sorry, I haven't seen this code.  I've been studying Azerothcore and the Sunwell project, but it does look very similar.  I'm currently researching ways to implement integrated NPCs, and the code I'm referencing is :
uint32 GetCurrentVendor() const { return m_currentVendorEntry; }
void SetCurrentVendor(uint32 vendorEntry) { m_currentVendorEntry = vendorEntry; }
    Rochet2 is a respected mentor, and I've used his AIO and Recasting.  This is my first time submitting, and if there's anything inappropriate or offensive, I can close this PR. Thank you.

**Comment by yangyitzzg** [comment]
> > > https://github.com/Rochet2/TrinityCore/compare/TrinityCore:3.3.5...multitrainer_3.3.5.diff
> > 
> > 
> > I'm sorry, I haven't seen this code. I've been studying Azerothcore and the Sunwell project, but it does look very similar. I'm currently researching ways to implement integrated NPCs, and the code I'm referencing is : uint32 GetCurrentVendor() const { return m_currentVendorEntry; } void SetCurrentVendor(uint32 vendorEntry) { m_currentVendorEntry = vendorEntry; } Rochet2 is a respected mentor, and I've used his AIO and Recasting. This is my first time submitting, and if there's anything inappropriate or offensive, I can close this PR. Thank you.
> 
> No, I don't have a problem with your PR, I put it here so that you can add the SQL part Like multi vendor it works with sql TOO

Sorry,  Please forgive a novice. Thank you.

File: src/server/game/Server/WorldSession.h [hunk 1]
```diff
@@ -367,6 +367,8 @@ class WorldSession
 
     uint32 GetCurrentVendor() const { return m_currentVendorEntry; }
     void SetCurrentVendor(uint32 vendorEntry) { m_currentVendorEntry = vendorEntry; }
+    uint32 GetCurrentTrainer() const { return m_currentTrainerEntry; }
```
Comment (Winfidonarleyan) [inline src/server/game/Server/WorldSession.h:370]:
```suggestion
    [[nodiscard]] uint32 GetCurrentTrainer() const { return m_currentTrainerEntry; }
```

**Comment by laasker** [comment]
> Add some content: creature_template:trainer_type Must be set to be greater than 3 and add UNIT_NPC_FLAG_TRAINER

I can't get it to work, are you sure it's trainer_type column that must be greater than 3?

Could you share the entry in creature_template of one of your npc trainers?

**Comment by yangyitzzg** [comment]
> > Add some content: creature_template:trainer_type Must be set to be greater than 3 and add UNIT_NPC_FLAG_TRAINER
> 
> I can't get it to work, are you sure it's trainer_type column that must be greater than 3?
> 
> Could you share the SQL insert from creature_template of one of your npc trainers?

Have you set up an NPC script: multiTrainerScript

**Comment by yangyitzzg** [comment]
`INSERT INTO `creature_template` (`entry`, `difficulty_entry_1`, `difficulty_entry_2`, `difficulty_entry_3`, `KillCredit1`, `KillCredit2`, `modelid1`, `modelid2`, `modelid3`, `modelid4`, `name`, `subname`, `IconName`, `gossip_menu_id`, `minlevel`, `maxlevel`, `exp`, `faction`, `npcflag`, `speed_walk`, `speed_run`, `speed_swim`, `speed_flight`, `detection_range`, `scale`, `rank`, `dmgschool`, `DamageModifier`, `BaseAttackTime`, `RangeAttackTime`, `BaseVariance`, `RangeVariance`, `unit_class`, `unit_flags`, `unit_flags2`, `dynamicflags`, `family`, `trainer_type`, `trainer_spell`, `trainer_class`, `trainer_race`, `type`, `type_flags`, `lootid`, `pickpocketloot`, `skinloot`, `PetSpellDataId`, `VehicleId`, `mingold`, `maxgold`, `AIName`, `MovementType`, `HoverHeight`, `HealthModifier`, `ManaModifier`, `ArmorModifier`, `ExperienceModifier`, `RacialLeader`, `movementId`, `RegenHealth`, `mechanic_immune_mask`, `spell_school_immune_mask`, `flags_extra`, `ScriptName`, `VerifiedBuild`) VALUES (600000, 0, 0, 0, 0, 0, 31833, 0, 0, 0, 'multTrainer', '', NULL, 0, 80, 80, 1, 35, 8177, 1, 1.14286, 1, 1, 20, 1, 0, 0, 1, 2000, 2000, 1, 1, 1, 2048, 2048, 0, 0, 4, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, '', 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 'multiTrainerScript', 0);
`

**Comment by laasker** [comment]
> > > Add some content: creature_template:trainer_type Must be set to be greater than 3 and add UNIT_NPC_FLAG_TRAINER
> > 
> > 
> > I can't get it to work, are you sure it's trainer_type column that must be greater than 3?
> > Could you share the SQL insert from creature_template of one of your npc trainers?
> 
> Have you set up an NPC script: multiTrainerScript

Yes, but i tested with the creature_template that you sent and it worked, thanks!

**Comment by sudlud** [comment]
what is the state of this PR? please address the review comment

**Comment by iThorgrim** [comment]
If you need to add support for SQL Gossip -> https://github.com/SoulStorm-Reforged/ForgeCore/commit/7f9a302c31c4a8b1d3c6c618ab4824ac03a4224b

**Comment by TheSCREWEDSoftware** [comment]
@yangyitzzg If you can provide the status of this PR

**Comment by Kitzunu** [comment]
this is abandoned. Feel free to reopen if you want to continue working on it

**Pull request closed without merging**